[14.0][ADD] crowdfunding

hbrunn opened the pull request:
This module provides the basics to turn Odoo into a crowdfunding
platform.

Configuration
=============

To configure this module, you need to:

1. Go to Crowdfunding/Settings
2. Review the preselected crowdfunding product that is used on vendor
   bills to pay people

Usage
=====

The module provides backend as well as frontend functionality.

For the backend:

1. Go to Crowdfunding
2. Create a new crowdfunding challenge
3. Set at least a name and a description
4. Click 'Open'

For the frontend:

1. Navigate to /crowdfunding on your Odoo website
2. Select a challenge from the list
3. Join the discussion, claim the challenge for yourself to solve or add
   your pledge

Known issues / Roadmap
======================

As this module is designed to provide the bare necessities, a couple of
other modules will add more advanced functionality

-  crowdfunding_membership allowing to restrict creating/claiming to
   members only
-  crowdfunding_project allowing to create crowdfunding challenges from
   projects/tasks or vice versa
-  crowdfunding_github allowing to create crowdfunding challenges from
   issues of some github project
-  crowdfunding_gitlab the same as above but for gitlab
-  crowdfunding_category allowing to move some of the base company wide
   configuration to categories

hbrunn committed:
[ADD] crowdfunding

vdewulf commented:
I tested the current work on runboat.

Here are my steps: 
- went to runboat build, chose the full database (not the base)
- go to crowdfunding / configuration and click on the Product "Crowdfunding". Changed the product type from consumable to Service (it makes more sense to me) and changed the public price to 100$
- still in the settings, changed the default fee percentage from 0 to 10%
- went to the Challenges Menu, Create a new one
- I updated the name, target amount. There is already 13 500$ on the Claimed Partner Amount, see screenshot. Why?
- 
<img width="1301" height="889" alt="Sélection_182" src="https://github.com/user-attachments/assets/96a20a08-d4ba-46a9-9ffe-4e50ec0d273f" />
- I added a Description image and a Description URL and a Description
- in the Miscellaneoouus tab, I activated "is published" button
- Saved then clicked on "Open" to change the status of the challenge.

Then I went on the website and saw the challenge: 
<img width="1154" height="815" alt="image" src="https://github.com/user-attachments/assets/b7b5531e-feda-4c64-b60d-00005f535af5" />
- clicked on "More Information" and this lead to the URL configured before
- the description, image and amount needed are displayed.
- There are 2 buttons : Claim and Pledge

Claim : I clicked on Claim button and now the Challenge is "claimed by Mitchell Admin". I suppose this means that the user is now the person who should do the work.
==> would this be configurable that not anyone can "claim" a challenge? I mean that we could decide that only maintainers or PSC's can Claim a challenge. How could we do that?

Pledge : clicking on Pledge button, this opens a new page where you can add any amount.
==> would it be possible to configure a default amount on each challenge, to encourage people to add at least a certain amount?

I validated an amount, and went to the Payment page. Chose Wire Transfer Mode.
The reference is displayed not super nicely: 
<img width="836" height="663" alt="image" src="https://github.com/user-attachments/assets/efe89f46-120c-4a06-bdbd-60116d28eb34" />

At this stage, when going back to the challenge webpage, nothing has changed, as the payment is not done yet.
On the Challenge in the backend, we see now there is one transaction: 
<img width="1540" height="385" alt="image" src="https://github.com/user-attachments/assets/42567166-61a2-4d33-b001-aee3c155c1d1" />

At this stage, I am blocked because the Transaction is standalone and I cannot edit its status. what would be the next steps for me to finalize the pledge?

Thanks a lot for this feature!

hbrunn commented:
thanks @vdewulf, I'm numbering replies for easier reference

1. the "claimed partner amount" is meant to be the amount that is paid to the partner (it's computed as target amount - fee amount) - I'll be happy to rename this to whatever you consider more intelligible
2. turning off claiming via some config option seems fine to me, actually, it might make more sense to not allow this at all in the base module and move this functionality to module `crowdfunding_public_claim`?
3. adding a default amount per challenge: sure, will implement soon
4. ugly reference: easy to fix display, do you have any preference for ie a running number instead of the guid?
5. handling payments: this relates to my previous question about when to create invoices, as there needs to be some accounting entry in the end for things to integrate smoothly with the rest of Odoo. The simple way of doing this is to create an invoice as soon as the user has submitted the amount to pay. Then I can just create the invoice, and send the user to the invoice payment workflow. Downside is that if a user then doesn't finish the payment, you have a draft invoice that needs further processing (either followup or delete). The other possibility is that I create the invoice only after a payment is finished, that will work smoothly for online payments, but for wire transfer you'll have the same issue that there are invoices that might never be paid. In both cases, an automatic reminder/cleanup cronjob could keep manual steps low here. From a reuse-as-much-existing-functionality-as-possible-perspective, creating the invoice immediately would have my preference.
6. I've clicked the 'set submitted' button in the backend on your challenge (that's what you do when the person supposed to work on this has handed in their results), in this state no further pledges are possible and a button shows up in the backend that allows you to create a vendor bill for the partner. Please check if this is the workflow you expect for dealing with paying people.

vdewulf commented:
Hi @hbrunn 

1. okay now I understand. Maybe "Net amount for the claimer"?
2.  agreed, better to move this feature to crowdfunding_public_claim
3.  default amount per challenge: thanks
4.  ugly reference: yes, I would prefer a sequence that includes the reference of the Challenge followed by a sequence that increases at each bidder instead of this long reference
5. When to create invoices: it seems that you didn't see my answers in the OCA Odoo Instance task. I agree with you, create the draft invoice right ahead in the process. Here is what I wrote a few days ago: 

" I think it is a good idea to get one invoice for each payment. It will help use the standard worfklow as you suggest, and as many people in the Odoo ecosystem are working for companies, I expect they will want to get an official document for their accounting.

If the invoices stay in draft if not paid in the end, it is not an issue as we can easily delete them on a regular basis (--> to be added in our internal process but nothing impossible to do!).

Currently we don't have a lot of draft invoices per se, because the process stays in the Sale Order step and Rebecca manually confirms the sales orders to create the invoices on a regular basis. This is something we want to improve after the migration to v18 so we might end up with more draft invoices that we will need to clean up.
Another information, might not be related but I prefer to communicate how we work: our Treasurer Frederik suggests us to use draft vendor bills in order to get the information of the committed expenses in MIS Reports. So in the beginning of the year and during the year, I create one draft invoice for my company for example, adding the budget of the year minus what was already invoiced. This is a bit boring to be honest because it requires an update of the draft invoice each time a invoice is received.

This might be useful to know because we won't be able to stupidly delete all the drafted invoices at once when cleaning regularly. I am personally in favor of finding another solution than the draft invoices in the future but in the meantime it will work this way."

6. Next steps of the process, once the "Set submitted" button was clicked by you, I could create an invoice with a % of the amount (this is nice to be able to pre-pay at the start, then during the work and keep the last bit in the end).

An improvement could be that the next time I hit the button to create an invoice, it doesn't allow me to set 100% again, limiting the risk of paying to much to the vendor. Nice to have: in the popup with the % to set, if the user could see an overview of the existing bills with the total amount billed for each vendor bill (with the % of the toal), it might help us know how much still needs to be invoiced.

Another question is the fact that no other pledge can be done once the first vendor bill has been created. This means that we will need to wait for a full 100% crowdfunding to launch the payement. I don't know if this bothers me right now. I think it is okay to try this way and update if we see it's blocking us. I was thinking about the OpenUpgrade scripts, it might be possible that we want to start paying the developers before we reach the total amount. But we could also just split the crowdfunding in different challenges and go level by level.

In the feedback received from the maintainers about the crowdfunding idea, many raised the issue about paying the reviewers or co-workers. In this case, could we have another module that allows to define more than one claimer and split the money between them in a free way (with % between them or by setting an amount).

That's it for now, but I'll need to review all the feedback from the maintainers to make sure to take them into account.

Thanks a lot!

hbrunn commented:
1. I made this just 'Amount payable' for brevity, and added a help text. Btw wherever you think help texts are appropriate, let me know, I think UIs should be as self-documenting as possible. Added a few placeholders too
2. done, I named the module crowdfunding_claim. You can uninstall it on runboat
3. added
4. done
5. done
6. vendor bill generation
    1. not allowing 100% again: for now, I set the percentage to 0 in case there are existing vendor bills. Reasoning for being a cautious with hard limits here is that you have a moving target - if there's more money pledged than the target amount, you'll want to pay that, correct? But still the wizard's percentage should be from the point of view of the target amount if the target is not reached yet, I think.
    2. showing the existing vendor bills there now
    3. you can't pledge any more when the challenge is in state 'submitted', so this has nothing to do with payments. Now I also show the create invoice button also in state 'claimed', makes more sense indeed
7. paying more than one partner: That indeed doesn't seem to me to belong into the base module. One of my customers has a nifty module that uses partner_multi_relation to define a partner that's actually a conglomerate of other partners, and using weights on the conglomerate relation splits invoices for a conglomerate partner into invoices for the partners constituting the conglomerate. That's on my list to OCAize for a long time already, and could be used to do this. That said, I'd rather expect that multiple parties work in a way we did with openupgrade, so that every party is paid for work done until the funding runs out. Let's stash this for now

My plan for the coming weeks before the OCA days is:

- make you happy
- write tests
- request at least one technical review
- merge this
- switch to an issue-base workflow, I already create issues meant to be picked up during the code sprint whenever I think of something that should be done here

vdewulf commented:
Hello @hbrunn 
I tested today on the runboat:
* the default percentage fees --> it works, I added in the settings then create a new challenge and it applied. I still can change it on the challenge itself.
* in the miscelanneous tab, I added a default pledge amount.
* I published and opened the challenge, went on the webiste and Pledged (the default amount is shown, I changed it). when submitting, I got this error:  
<img width="1348" height="353" alt="image" src="https://github.com/user-attachments/assets/6b4890cb-e77b-44d6-ba69-a62b3fe5359e" />

So I stopped my testing of the next steps in this flow.

Test of vendor billing:
* I then removed the claim module on the runboat. The "claim" button on the website page is not displayed anymore. I could add a claimer in the backend on the form of the challenge.
* I cicked on the button "Set Claimed" ==> what's the impact of this status "Claimed"?
* I then clicked on the button "Set submitted" ==> I see that now it is not possible to pledge the challenge on the website anymore. So this should be clicked when we got enough money and we don't want to get more money. Is it right? Could be nice to add help texts on the status.
* Created a first invoice of 50%. I'm wondering about the VAT included or excluded. It seems that the target is VAT excluded, but the invoicing takes into account the VAT. So is this consistent? Maybe we should add a help text on the Target amount to specify if it is including VAT or not?
* The display of already invoiced amount works fine! What misses maybe is the initial % choosen at that time, this could help decide how much % to add in the popup based on the previous % invoiced.
* A nice to have maybe: in the challenge, below the Amount payable field, add a field "invoiced amount" so that we have a clear view of what is the total we still need to receive from the claimers.
* In the list view, we could also add the field "amount payable" and "invoiced amount" to get an overview

Feedback regarding the multiple claimers question: for OpenUpgrade until now you each invoiced what you did, and it is not a % between you. So how would this work in the crowdfunding system in the future? I don't see it clearly from an administrative perspective. But we can discuss it in September together.

Let me know when I can test the pledge part with the invoicing aspects for the pledgers.

Thanks a lot!

hbrunn commented:
>     * I published and opened the challenge, went on the webiste and Pledged (the default amount is shown, I changed it). when submitting, I got this error:

that's fixed now

>     * I cicked on the button "Set Claimed" ==> what's the impact of this status "Claimed"?

nothing, but I think this state is useful for filtering

>     * I then clicked on the button "Set submitted" ==> I see that now it is not possible to pledge the challenge on the website anymore. So this should be clicked when we got enough money and we don't want to get more money. Is it right? Could be nice to add help texts on the status.

done

>     * Created a first invoice of 50%. I'm wondering about the VAT included or excluded. It seems that the target is VAT excluded, but the invoicing takes into account the VAT. So is this consistent? Maybe we should add a help text on the Target amount to specify if it is including VAT or not?

taxes depend entirely on the configuration of the crowdfunding product and the default taxes (General Settings/Invoicing), typically you'll have the default taxes disabled in a real world Odoo installation, and handle this on the product.

Actually it's unclear to me what the taxing situation here is at all, will this be considered tax free donations? And for work done, non-CH companies will be exempted from VAT I guess, and Swiss contributors just have to suck it up that they'll have an included tax?

But indeed, it's more logical to sum up the total amount there, which I do now

>     * The display of already invoiced amount works fine! What misses maybe is the initial % choosen at that time, this could help decide how much % to add in the popup based on the previous % invoiced.

I show the percentage based on already paid vendor bills, or would you rather count draft/unpaid vendor bills too?

>     * A nice to have maybe: in the challenge, below the Amount payable field, add a field "invoiced amount" so that we have a clear view of what is the total we still need to receive from the claimers.

I called this "unpaid pledged", because with "invoiced" it's unclear if this is about vendor bills or customer invoices, and with crowdfunding we handle both

>     * In the list view, we could also add the field "amount payable" and "invoiced amount" to get an overview

added that and a couple more

> Feedback regarding the multiple claimers question: for OpenUpgrade until now you each invoiced what you did, and it is not a % between you. So how would this work in the crowdfunding system in the future? I don't see it clearly from an administrative perspective. But we can discuss it in September together.

The mechanics already allow that you attach an arbitrary vendor bill to a challenge. Is it sufficient for you if this then also shows the amount pledged and the amount already paid on a draft invoice when you select a challenge there?

vdewulf commented:
Hi @hbrunn 
Thanks for your replies and bug fixes.
The remaining comments from my side after testing today: 
* The popup to invoice to the vendor seemed bugged. It indicates that the invoice is "paid" but it is actually not: 
<img width="1183" height="561" alt="image" src="https://github.com/user-attachments/assets/40a2c7f2-4fc1-440c-b8b7-1c5e55fee0b7" />
Invoice:
<img width="1854" height="530" alt="image" src="https://github.com/user-attachments/assets/ac74b4da-15cd-44d5-86f9-e583e1a66474" />

* Regarding the % of already invoiced (or paid), I think it is meaningful to take into account the invoiced in status "posted" (not draft, not cancelled), whether they are paid or not yet. It is possible that you need to receive invoices quickly twice in a small period of time and you haven't done your bank reconciliation yet. So one of the invoice is not "paid" in Odoo yet but is in reality already paid.

* I invoiced another vendor bill through the same challenge. When trying to confirm it, I got this error regarding a duplicate of vendor bill reference. In reality, we will receive the real invoice from the vendor and we'll need to input their reference in this field. So it won't provoke any issue. However for testing purpose, might be good to add a number after the reference or something that makes it differnt from the others?
<img width="1425" height="491" alt="image" src="https://github.com/user-attachments/assets/0592a82a-8399-4996-a754-d163f67d3f39" />
 
When I added a manual payment to one of both invoices, the field "Amount paid out" was updated correctly on the challenge.

* Regarding the mechanics of adding any challenge to a Vendor Bill, I noticed it now and it's great! I just wonder what's the link with the "Claimed by" field.
And also I could not see the field "Crowdfunding Challenge" on a draft Bill Invoice: 
<img width="952" height="633" alt="image" src="https://github.com/user-attachments/assets/f24af340-faf1-42f4-bb1c-20224f0ae79d" />

Thanks!

hbrunn commented:
>     * The popup to invoice to the vendor seemed bugged. It indicates that the invoice is "paid" but it is actually not:

that's a weird bug in v14, worked around that

>     * Regarding the % of already invoiced (or paid), I think it is meaningful to take into account the invoiced in status "posted" (not draft, not cancelled), whether they are paid or not yet. It is possible that you need to receive invoices quickly twice in a small period of time and you haven't done your bank reconciliation yet. So one of the invoice is not "paid" in Odoo yet but is in reality already paid.

changed the relevant selections to state == 'posted'

>     * I invoiced another vendor bill through the same challenge. When trying to confirm it, I got this error regarding a duplicate of vendor bill reference. In reality, we will receive the real invoice from the vendor and we'll need to input their reference in this field. So it won't provoke any issue. However for testing purpose, might be good to add a number after the reference or something that makes it differnt from the others?

then I just keep the reference empty

>     * Regarding the mechanics of adding any challenge to a Vendor Bill, I noticed it now and it's great! I just wonder what's the link with the "Claimed by" field.
>       And also I could not see the field "Crowdfunding Challenge" on a draft Bill Invoice:

original plan was to show it only when it's nonempty, but makes sense to show it in draft state too. I added some informational fields there too when you select a challenge, should help with cost control. There's no relation with the claimed by field, should there be one?

vdewulf commented:
Hello @hbrunn 
Thanks for the last changes. I won't be able to test them before the OCA Days. @TumbaoJu will take over and get familiarized with the module before the OCA Days. She will be available on the Monday for the fine-tuning tests.
See you there!

alessandrograndi commented:
Hi! I functionally tried the module and it seems to work fine.

I just found out a couple minor issues that can improve user experience.

1) When the portal user claims the task the page is redirected to a "<path>/claim" url that doesn't really exist after the claiming. So if the user tries to add a comment, a 404 error will be given (very edge case I think)

2) When creating the Vendor Bill the `Percentage` field will compute the invoce value **without taxes**. If you set it to 100%, the value of the invoice will be higher than the payable amount and preventing from setting the Challenge to `Done`.
<img width="925" height="193" alt="immagine" src="https://github.com/user-attachments/assets/0b46aa3a-9dd4-4f22-a9b5-9ea950d7b73e" />

This was a little bit confusing to me also because the `Percentage Paid` seems to be taking in account taxes on the bills already invoiced.
<img width="971" height="257" alt="immagine" src="https://github.com/user-attachments/assets/a9cc440d-869b-404e-ae95-68da6d48eec2" />

I'm not sure if the `claimed_partner_amount` is intended to be tax included or not but (if that is the case) it would be nice to have the parcentage field taking the tax in consideration too (in order to avoid manual computation).

houssine78 commented on crowdfunding/models/crowdfunding_challenge.py:
```diff
@@ -0,0 +1,342 @@
+            this.pledged_amount_unpaid = (
+                sum(this.invoice_ids.mapped("amount_total")) - this.pledged_amount
+            )
+            this.pledged_amount_total = this.pledged_amount + this.pledged_amount_unpaid
```
do we need to take care of the case where for any reason a validated invoice has been reversed totally or partially in a credit note  or it's out of scope ? actually the filter to out_invoice state doesn't allow to take credit note into account for the computation of the different amount.

Not blocking for this first version but maybe worse mentioning it in the limitation section.

Same goes for the vendor bills

hbrunn replied:
good point. do you agree just changing the domains (adding out_refund/in_refund there) in the one2many fields will do the trick?

At the same time, we probably also carry the link to the crowdfunding if any to reversed moves.

houssine78 replied:
I was wandering if the calculation needs to switch to signed amounts but after checking the code it seems that it will work with just changing the domain the way you suggested.

davejames commented:
Functional and technical review. I went through the whole workflow, I have proposed a PR to your fork with some performance and UX suggestions. Everything works, and looks good to me, but I would suggest the invoicing wizard button use "vendor bill" terminology, as the invoices are "pledge" amounts, while the wizard creates vendor bills for the "claimed by" partner.

hbrunn commented:
@alessandrograndi thanks for the input:

1. fixed that by not showing the chatter in that case
2. that's a result of the runboat installation setting default taxes which are exclusive. I amended crowdfunding_demo to remove default taxes (so that testers don't have to remember removing them), an

houssine78 commented on crowdfunding/readme/CONTRIBUTORS.md:
```diff
@@ -0,0 +1 @@
+-   Holger Brunn \<mail@hunki-enterprises.com> (https://hunki-enterprises.com)
```
```suggestion
-   Holger Brunn \<mail@hunki-enterprises.com> (https://hunki-enterprises.com)
-   Houssine Bakkali \<houssine.bakkali@gmail.com> Open Architects Consulting SRL (https://github.com/houssine78)
```

norlinhenrik commented on crowdfunding/readme/CONTRIBUTORS.md:
```diff
@@ -0,0 +1,2 @@
+-   Holger Brunn \<mail@hunki-enterprises.com> (https://hunki-enterprises.com)
+-   Houssine Bakkali \<houssine.bakkali@gmail.com> Open Architects Consulting SRL (https://github.com/houssine78)
```
```suggestion
-   Houssine Bakkali \<houssine.bakkali@gmail.com> Open Architects Consulting SRL (https://github.com/houssine78)
-   Henrik Norlin \<henrik@loym.com>
```

houssine78 reviewed (approved):
LGTM

hbrunn commented:
/ocabot merge nobump
